File: Go/sort/quick_sort.go
package sort

func QuickSort(arr []int) {
	Partition(arr, 0, len(arr)-1)
}

func Partition(arr []int, left int, right int) {
	if left < right {
		peek := Sort(arr, left, right)
		Partition(arr, left, peek-1)
		Partition(arr, peek+1, right)
	}
}

func Sort(arr []int, left int, right int) int {
	pivot := arr[left]
	low := left + 1
	high := right

	for {
		for low <= high && arr[low] <= pivot {
			low++
		}
		for low <= high && arr[high] >= pivot {
			high--
		}
		if low > high {
			break
		}
		arr[low], arr[high] = arr[high], arr[low]
	}

	arr[left], arr[high] = arr[high], arr[left]
	return high
}

File: Python/sort/QuickSort.py
class QuickSort:
    def __init__(self, arr):
        self.arr = arr
        self.length = len(arr)
        self.quicksort(0, self.length - 1)

    def partition(self, left, right):
        pivot = self.arr[left]
        low = left
        high = right
        while low < high:
            low += 1
            while low <= right and self.arr[low] <= pivot:
                low += 1
            while self.arr[high] > pivot:
                high -= 1
            if low < high:
                self.arr[low], self.arr[high] = self.arr[high], self.arr[low]
        self.arr[left], self.arr[high] = self.arr[high], self.arr[left]
        return high

    def quicksort(self, start, end):
        if start < end:
            index = self.partition(start, end)
            self.quicksort(start, index - 1)
            self.quicksort(index + 1, end)